### ISSUE TITLE
unable to replicate a result from Athalye et al 2018 with art -- issue with gradient calculation?

### ISSUE THREAD

**Issue by Embeddave** [Trusted-AI/adversarial-robustness-toolbox#702]
Hi `art` devs

This isn't exactly a bug, but it's an issue we've been fighting with that's relevant for our research under GARD, and we've been stuck on it long enough we thought it was time to reach out for some help.

Apologies in advance for the lengthy issue! But we felt this was worth discussing further. Any insight you can give us on where we might be going wrong would be much appreciated.

**Describe the bug**

We're unable to use `art` to successfully replicate a results from Athalye Carlini Wagner 2018, reporting successful attack of the JPEG compression defense, using BPDA.

Hope it's ok if I tag in some people at two-six labs we have been in touch with about our defense. We thought you might want to be aware of this since you are using JPEG compression in some example `armory` scenario configs.
cc @davidslater @ng390 @kevinmerchant

Some context: we want to rigorously attack our own defense, which is stochastic, but so far we have been unable to successfully do so with any version of Carlini-Wagner in `art`. We are testing an implementation of our defense that subclasses `art.Preprocessor`, including an `estimate_gradient` method that simply returns the passed-in `grad`, i.e. performs an implicit identity transform on the backward pass.

As a "positive control", to make sure we *can* successfully attack __some__ defense, we chose JPEG compression, as attacked in [Athalye Carlini Wagner 2018](https://arxiv.org/abs/1802.00420).

I'm guessing you know the paper, but just to have all the info in place on this issue:  
Our understanding is they used PGD-Linf with BPDA, epsilon 0.05, where they substituted the identity transform for JPEG compression on the backward pass, under the assumption that g(x) ~ x, where g(x) is the JPEG transform. From their paper: 
> "To circumvent bit-depth reduction and JPEG compression, we use BPDA and approximate the backward pass with the identity function. ... With our attack, we achieve 100% targeted attack success rate and accuracy drops to 0% for the strongest defense under the smallest perturbation budget considered in Guo et al. (2018), a root-mean-square perturbation of 0.05 (and a “normalized” l2 perturbation as defined in Guo et al. (2018) of 0.01)."

Their demo is here.
https://github.com/anishathalye/obfuscated-gradients/blob/master/inputtransformations/jpeg.ipynb
The comments + defaults support the idea they used PGD and JPEG quality of 75% as reported in the Guo et al paper they refer to. Unfortunately we were not able to figure out how to install their code so that we could produce an "oracle" to test against--unclear which version(s) of Tensorflow they used?

Using `art`, we fail to replicate their key result.
We use an `art` attack given a `PyTorchClassifier` that uses `JPEG_compression` as a `preprocessing_defense`, with the `quality` set to `75`. Even when we go up to 10k iterations, we achieve an 88% success rate, not 100% as they report. See screenshot below (code attached to reproduce).
![Screen Shot 2020-10-29 at 9 40 02 PM](https://user-images.githubusercontent.com/58785101/97757147-8b73ab80-1ad2-11eb-9890-1e5d7d47e846.png)

Note we are attacking a `resnet50` model trained on CIFAR10 with the Madry lab `robustness` library, but __without__ adversarial training. The model checkpoint is here: https://www.dropbox.com/s/yhpp4yws7sgi6lj/cifar_nat.pt?dl=0

As a sanity check we set JPEG quality to 100%. In that case we did find we were able to achieve a success rate of 100%. However we found that this success rate quickly drops off as we lower the quality of JPEG compression. See screenshot of plot below.

![Screen Shot 2020-10-30 at 5 08 49 PM](https://user-images.githubusercontent.com/58785101/97757182-9f1f1200-1ad2-11eb-924c-79e24199c66c.png)

So at this point it is unclear to us whether:
(1) we are failing to use `art` properly to replicate the approach of Athalye Carlini Wagner 2018
(2) some sort of numerical error is impairing but not completely breaking the gradient estimation
or alternatively (3) the defense is not as easily broken as the paper suggests, or there is some detail in the paper that we have not understood / has been left out.

WRT (2) above, we do note that we see warnings about low-level computations related to the gradient when the `art` attack is operating on the classifier *with* the pre-processing defense. These are the errors we see: 

```
JPEG compression: 64it [00:00, 3642.03it/s]
[W accumulate_grad.h:165] Warning: grad and param do not obey the gradient layout contract. This is not an error, but may impair perfor
mance.
grad.sizes() = [2048, 512, 1, 1], strides() = [512, 1, 1, 1]
param.sizes() = [2048, 512, 1, 1], strides() = [512, 1, 512, 512] (function operator())
JPEG compression: 64it [00:00, 3621.49it/s]
JPEG compression: 64it [00:00, 3599.10it/s]
...
batch: 1
[W TensorIterator.cpp:918] Warning: Mixed memory format inputs detected while calling the operator. The operator will output contiguous tensor even if some of the inputs are in channels_last format. (function operator())
JPEG compression: 64it [00:00, 3605.29it/s]
```

This suggests the error could arise from some complex interaction between PyTorch C backend + CUDA + art.
The only way we can think of to confirm this now would be to implement the attack including BPDA in pure PyTorch and test whether we can then successfully circumvent JPEG compression.

**To Reproduce**
I am attaching a notebook that shows what we did to test JPEG Compression with quality set to 75%.
I am also attaching a complete script using `art` that shows our attempt to measure how quickly the success rate drops off, and a notebook to visualize the results from that script.
Our results are included as a .csv that is visualized with that notebook.
`conda` environment files are also attached.
[Archive.zip](https://github.com/Trusted-AI/adversarial-robustness-toolbox/files/5467674/Archive.zip)

**Expected behavior**
Attack should be 100% successful against  when using BPDA.
Our understanding is that the `estimate_gradient` method of the `JPEGCompression` class provides this.

**System information (please complete the following information):**
 - OS: Ubuntu 18 -- default AWS EC2 instance install
 - Python 3.6
 - ART 1.4.1
 - PyTorch 1.6

please use attached environment.yml and spec-file.txt to exactly reproduce our environment (in the archive.zip)

**Comment by beat-buesser** [comment]
Hi @Embeddave First, I have to say thank you very much for taking the time to write this comprehensive description of your experiments, thoughts and insight and for providing example scripts and models to reproduce them, this is really exceptional!

We are very interested to learn what's causing this and our team will take a closer look as soon as possible and come back with either questions to get a better understanding or a solution very soon.

**Comment by Embeddave** [comment]
Thank you @beat-buesser 
I have empathy for OS devs :)
Look forward to working together on this more

**Comment by beat-buesser** [comment]
Hi @Embeddave 

I have been able to run your notebook, which was straightforward with the great material that you have provided. I also have observed similar benign and adversarial accuracy values as you are describing.

I have not noticed any warnings related to low-level computations running the notebook on a V100 and I'm wondering if you are running any mixed precision calculations?

I have noticed that you are running the attacks with `attack.generate(x=x)` which means that the attack tries to change the predicted label, but afterwards you calculate the accuracy with the true labels `y`. In this kind of attack on predicted labels it can happen that  predictions of wrong classes can get changed by adversarial perturbation into the correct class of the true label `y` which then prevents the accuracy calculated with true labels from becoming 0%.

Providing the true labels `y` to the attack with `attack.generate(x=x, y=y)` makes the attack stronger against the true labels because all changed predicted labels will be away from the true label. With these changes the notebook achieves 0% accuracy with and without defence, often after only 10 iterations.

**Comment by davidslater** [comment]
The other things I would add to Beat are:
1) The Athalye paper didn't mention many of the attack parameters (PGD? Number of iterations? JPEG compression parameter?), so you may need to modify some of those things, like random restarts.
2) The metric used is *not* Linf, it is L2: "perturbation budget considered in Guo et al. (2018), a root-mean-square perturbation of 0.05 (and a “normalized” l2 perturbation as defined in Guo et al. (2018) of 0.01)" [https://arxiv.org/pdf/1711.00117.pdf (Guo paper)]. RMS is the square root of L2, so they are looking at an L2 of 0.0025 if I'm reading that correctly. The normalized L2 perturbation you would need to look at Guo.
3) As the JPEG compression becomes more lossy, identity BPDA becomes an increasingly worse approximation. I would think that it would have a hard time after a while.

**Comment by Embeddave** [comment]
Thank you @beat-buesser your feedback is quite helpful.

> I have noticed that you are running the attacks with `attack.generate(x=x)` which means that the attack tries to change the predicted label, but afterwards you calculate the accuracy with the true labels `y`. In this kind of attack on predicted labels it can happen that predictions of wrong classes can get changed by adversarial perturbation into the correct class of the true label `y` which then prevents the accuracy calculated with true labels from becoming 0%.
> 
> Providing the true labels `y` to the attack with `attack.generate(x=x, y=y)` makes the attack stronger against the true labels because all changed predicted labels will be away from the true label. With these changes the notebook achieves 0% accuracy with and without defence, often after only 10 iterations.

I included the ground truth labels `y` and was also able to achieve 100% success rate with 10 iterations, given the other parameters I had in that notebook. My fault for not grasping the nuances of white box setting.

I am still not sure that this exactly replicates the method of Athalya et al. -- will explain in a reply below

As for the low-level error messages:  
> I have not noticed any warnings related to low-level computations running the notebook on a V100 and I'm wondering if you are running any mixed precision calculations?

I am not running mixed-precision calculations AFAIK.
This issue on the PyTorch repo suggests multiple people see similar errors for a variety of reasons:  
https://github.com/pytorch/pytorch/issues/42300
including training on one device then loading the same checkpoint on another device, wonder if that's what's happening here

**Comment by Embeddave** [comment]
Thank you @davidslater 

Just for future reference if anyone finds this issue:  

>     1. The Athalye paper didn't mention many of the attack parameters (PGD? Number of iterations? JPEG compression parameter?), so you may need to modify some of those things, like random restarts.

* yes you are totally right that the attack specifies L2, what they call "root-mean-square".
* definitely unclear which attack they used, since they only state two: PGD for L-Inf bounded, and CW for L2 perturbation, but of course this is a *bounded* L2 attack since they specify epsilon. Comments in their demo notebook suggest it was PGD. https://github.com/anishathalye/obfuscated-gradients/blob/master/inputtransformations/jpeg.ipynb
* Unclear to us whether their "RMS" is "just" L2, looks like they normalize by dividing by the number of pixels per image? See again in that demo notebook
* They use the same Jpeg quality as Guo et al, 75. I confirmed this in the `obfuscated-gradients` repo: https://github.com/anishathalye/obfuscated-gradients/blob/72e24bc4f4669c01a4f2987a2c4a01d2ba363687/inputtransformations/defense.py#L17

After making the fix that @beat-buesser suggested, I was able to successfully attack JPEG as a preprocessor (i.e., with "implicit" BPDA). 
- For PGD-L2 I can successfully attack using eps of 0.0025 and 1k iters, all other hyperparams are default.
- For CW L2 using all default hyperparameters, except `max_iter` which I lowered to 5. 
Did not measure the mean perturbation though.

So at least we have a "positive control" where we know we can successfully attack.

Thanks again for your help.
I will close this issue.
